### ISSUE TITLE
check-script-has-no-table-name gives false errors with certain keywords

### ISSUE THREAD

**Issue by jmriego** [Montreal-Analytics/dbt-gloss#36]
**Describe the bug**
check-script-has-no-table-name gives false errors on places where SQL uses `FROM` but it's not a table.
For example BigQuery has `EXTRACT(date FROM datetime_column)`

**To Reproduce**
```
WITH date_spine AS (
  {{ dbt_utils.date_spine(
      start_date="cast(parse_date('%Y/%m/%d', '2014/11/01') AS datetime)",
      datepart="day",
      end_date="cast(date_add(current_date(), INTERVAL 40 YEAR) AS datetime)"
     )
  }}
)

SELECT
    EXTRACT(YEAR FROM date_day) AS year,
    EXTRACT(QUARTER FROM date_day) AS quarter,
    EXTRACT(MONTH FROM date_day) AS month,
    EXTRACT(DAY FROM date_day) AS day

FROM date_spine
```
Run the check-script-has-no-table-name and it will fail with:
```
models/reference/calendar.sql: does not use source() or ref() macros for tables:
- date_day
```

**Expected behavior**
I would expect this to be ignored

**Version:**
v1.0.0

**Additional context**
Nothing else I can think of

**Comment by cnolanminich** [comment]
I think this is a feature / bug of using regex to parse the SQL, vs. a SQL parser.

`sqlfluff` has a parser that can extract only the tables (and, crucially, _not_ the CTEs), which I believe could be used as a drop-in replacement for [has_table_name()`](https://github.com/offbi/pre-commit-dbt/blob/main/pre_commit_dbt/check_script_has_no_table_name.py#L44-L69), defined something like this like this:

```python
import sqlfluff

def has_table_name(
    sql: str, filename: str, dotless: Optional[bool] = False
) -> Tuple[int, Set[str]]:
    # not sure if this status code would still be necessary
    status_code = 0
    parsed_sql = sqlfluff.parse(sql)
    table_names = set(parsed_sql .tree.get_table_references())
    return status_code, table_names
```

The downside to this is, this would tie this project to the `sqlfluff` dependencies, and the python API isn't quite set. The upside is that the code would allow for a more robust parser (regex on SQL to understand the SQL parse tree will always be an 80% solution, which is problematic is a CI setting).

https://github.com/sqlfluff/sqlfluff/blob/main/examples/03_extracting_references.py#L19-L20

**Comment by jmriego** [comment]
I totally agree with that and I understand sqlfluff is not ready yet for something like this. But wouldn't it be possible to modify the regex so it ignores things like `extract(month|year|... from)`?
That would at least add one extra case covered that seems to be used in several databases

**Comment by eugenekim-orrum** [comment]
I have this same issue with date_spine.

**Comment by Sebastian-Trustpower** [comment]
Similar problem

`SELECT 'whatever you want to type here from something' AS message`

will return an error 
**does not use source() or ref() macros for tables:** 
because it use the part of text as a table reference

**Comment by Subhabrata1996** [comment]
I am facing something similar with UNNEST in bigquery. example SQL :

`SELECT  * FROM UNNEST([1,2,3,4]) AS NUMBERS`

**Comment by i-zaykov** [comment]
same issue here with bq

**Comment by lidalei** [comment]
In Redshift / Postgres, `select substring(col FROM POSITION('a' IN col) + 1 for 10)` also gives an error.

**Comment by wilson-urdaneta** [comment]
Same when doing something like `trim(device.web_info.hostname::varchar(100))as device_web_info_hostname,`

Pretty much any unnesting logic

**Comment by tommyh** [comment]
@tomsej @BAntonellini - as a workaround, would it be possible to add an `--exclude` or `--ignore` flag, so we can configure dbt-checkpoint to ignore specific edgecases?  (While I agree that this isn't *optimal* and the list will grow over time, the specific number of offending keywords we have are quite low.)

**Comment by tommyh** [comment]
We hit an interesting edge case, which can be added here (string parsing vs sql parsing): sql string which contains the keyword `from`.

example:

```
select
  CASE
    WHEN 'hello' then 'foo from bar'
    ELSE 'goodbye'
   END as name
from {{ ref('stg_users') }}
```

will give this error:

```
models/my_model.sql: does not use source() or ref() macros for tables:
 - bar'
```

It sees `'foo from bar'` and thinks `from bar` is a table usage.  One interesting thing to note: it is parsing the table name as `bar'`, and not `bar`.

**Comment by SBSDataLabs** [comment]
similar i guess to @tommyh. i have a table with a column called "Created From Nonstock Item" which triggers a fail

Obviously a horrible name, but in this particular scenario out of my control.

**Comment by cmardiros** [comment]
This also fails in case statements that use `is distinct from` comparison operator in BigQuery at least. It seems to expect something to follow `from`. Given that case statements come before the model reference, the parser probably chokes on that first (using `is distinct from` in a where clause that comes after the model reference does not seem to pose a problem)

**Comment by vapolonio** [comment]
I have the same issue here using the unnest command with bigquery.
Maybe we can add some parameter to skip the checks on a specific set of models. If this is a valid solution, I can make the changes in the code.

**Comment by tayomoore** [comment]
Also ran into this issue today in this scenario
```
        , SUM(
            CASE
                WHEN field IN (
                    'foo from bar'
                    , '...'
                    , '...'
                ) THEN table.field
                ELSE 0
            END
        )
```
the `foo from bar` is being flagged

**Comment by janaperedni** [comment]
The problem still persists...

**Comment by damienchambon** [comment]
Any news on this? I think the PR made by janposlusny would fix it. Thanks.

**Comment by damienchambon** [comment]
By the way, I was able to bypass the bug by passing the arg `--ignore-dotless-table` to the command.

**Comment by BAntonellini** [comment]
#291 has been merged.

We are working on reviewing and introducing changes, preparing for a `v2.0.7` release. 

In the meantime, you can point to `rev: 955881d`

**Comment by chelsea-sezzle** [comment]
We tested with `rev: 955881d` — it fixed many false positives, thank you for the improvements!

However, we’re still seeing cases like the following being flagged:
```
SELECT
  qpReferral.value AS url_parameter_referral
FROM
  {{ source("some_schema", "some_table") }} AS t
  LEFT JOIN t.query_params qpReferral ON qpReferral.key = 'referral'
```
Hook output:
```
models/example_model.sql: does not use source() or ref() macros for tables:
 - t.query_params
```
Here, `t.query_params` is a nested field within the base table, not a separate relation—so using `ref()` or `source()` doesn’t apply.

Would be great if the check could exclude self-joins on nested fields like this. Happy to share more examples if helpful. Thanks again!

**Comment by martin-gonzalez-clarity** [comment]
Also tried with `rev: 955881d`/2.0.7 tag and unfortunately using Snowflake's `flatten` it also fails.
